_posts/2021-03-15-service-provider-interface.md

```diff
@@ -0,0 +1,201 @@
+
+We note the same thing with the MySQL driver: The `java.sql.Driver` located in `META-INF/services` contains `com.mysql.cj.jdbc.Driver` which is the implementation of the `Driver` interface of MySQL database.
+If the two drivers are loaded in the classpath, the `ServiceLoader` will read the implementation class names from each file, then calls `Class.forName` with the class names and then `newInstance` to create an instance of the implementation classes.
+
+### Example of an SPI Implementation
+Now that we have an understanding of the SPI concepts, let's create an example of an SPI and load providers using the `ServiceLoader` class.
+Let's say that we have a librarian who needs an application to check whether a book is available in his library or not when requested by customers.
+We can do this by defining a service represented by a class named `LibraryService` and a service provider interface called `Library`.
+The `LibraryService` provides a singleton `LibraryService` object. This object retrieves the book from `Library` providers.
+The library service client which is in our case the application that we are building gets an instance of this service, and the service will search, instantiate and use `Library` service providers.
+The application developers may in the first place use a standard list of books that can be available in all libraries. Another user who deals with computer science books may require another list of books for his library (another library provider). In this case, it would be better if the user can add the new library with the desired books to the existing application without modifying its core functionality. The new library will just be plugged into the application.
+
+### Implementation
+We start by creating a `maven` root project that will contain all our sub-modules. We will call it `library-service-provider-example`.
+The sub-modules will be:
+- *library-service-provider*: Contains the Service Provider Interface `Library` and the service class to load the providers.
+- *library-demo*: An application to put all together and create a working example.
+- *standard-library*: The provider for a standard library chosen by the developers.
+- *custom-library*: The provider for a custom library required by the user who deals with computer science books.
+
+#### library-service-provider
+
+First, let us create a model class that represents a book:
+```java
+public class Book {
+  String name;
+  String author;
+  String description;
+}
+```
+Then, we define an SPI for our service:
+```java
+public interface Library {
+  Book getBook(String name);
+}
+```
+Finally, we create the `LibraryService` class that the client will use to get the books from the library.
+```java
+private static LibraryService libraryService;  
+private final ServiceLoader<Library> loader;  
+
```

Comment:
I added the `getInstance()` method here because otherwise, the readers will ask how to get an instance. I also included the class name to make the code example complete.